illumos-utils/src/host/byte_queue.rs

```diff
@@ -0,0 +1,37 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::VecDeque;
+use std::sync::{Arc, Mutex};
+
+/// A queue of bytes that can selectively act as a reader or writer,
+/// which can also be cloned.
+///
+/// This is primarily used to emulate stdin / stdout / stderr.
+#[derive(Clone)]
+pub(crate) struct ByteQueue {
```

Comment:
Got it. I still think the name is misleading - nothing about `ByteQueue` or `byte_queue.clone()` implies that cloning will _not_ give you a copy of the byte queue that you now own, but instead will give you a shared reference to the same underlying `VecDeque`. This might be overly nitpicky because this is only used in the fake case, but maybe `SharedByteQueue` would be clearer? Or we could even say `SharedVecDeque<T>` and `impl io::{Read,Write} for SharedVecDeque<u8>` to be very explicit about what this thing really is.

Reply:
Renamed to `SharedByteQueue`!